File: leetcode/count-binary-substrings.cpp
// https://leetcode.com/problems/count-binary-substrings/

class Solution {
public:
    int countBinarySubstrings(string s) {
        int ans = 0;
        int i = 0;
        int n = 0;
        int k = s.size();
        while (i < k) {
            int m = 0;
            while (s[i] == s[i + m]) m += 1;
            ans += min(n, m);
            n = m;
            i += m;
        }
        return ans;
    }
};

File: leetcode/count-binary-substrings.hs
-- https://leetcode.com/problems/count-binary-substrings/

splitByEqualRuns :: Eq a => [a] -> [[a]]
splitByEqualRuns [] = []
splitByEqualRuns (x:xs) =
    let (as, bs) = span (== x) xs
    in (x:as) : splitByEqualRuns bs


pairwiseWith :: (a -> a -> b) -> [a] -> [b]
pairwiseWith f xs = zipWith f xs (tail xs)


countBinarySubstrings :: Eq a => [a] -> Int
countBinarySubstrings =
    sum . pairwiseWith min . map length . splitByEqualRuns
